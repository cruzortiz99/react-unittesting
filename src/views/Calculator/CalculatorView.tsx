import React from "react";
import styles from "./CalculatorView.module.css";

export enum CalculatorOperation {
  ADD = "add",
  SUB = "sub",
  MULTIPLY = "multiply",
  DIVISION = "div",
}

export function getCalculatorOperationFromString(value: string): CalculatorOperation {
  return Object.entries(CalculatorOperation)
    .filter(
      ([_, operationValue]) =>
        operationValue.toLowerCase() === value.toLowerCase()
    )
    .reduce<CalculatorOperation>((acc, current) => current[1] || acc, CalculatorOperation.ADD);
}

export interface ICalculatorView {
  firstNumber: number;
  onFirstNumberChange?: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  secondNumber: number;
  onSecondNumberChange?: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement> 
  ) => void;
  result: number;
  hasError?: string;
  operation: CalculatorOperation;
  onOperationSelected?: (event: CalculatorOperation) => void;
}

export function CalcultatorView(props: ICalculatorView): JSX.Element {
  return (
    <div>
      <h3>Calculator</h3>
      <div className={styles.calculatorContainer}>
        <input
          data-testid="firstNumber"
          type="number"
          placeholder="Number1"
          value={props.firstNumber}
          onChange={
            props.onFirstNumberChange ? props.onFirstNumberChange : () => {}
          }
        />
        <select
          data-testid="operation"
          value={props.operation}
          onChange={(event) =>
            props.onOperationSelected
              ? props.onOperationSelected(
                  getCalculatorOperationFromString(event.target.value)
                )
              : null
          }
        >
          <option data-testid="option-add" value={CalculatorOperation.ADD}>
            +
          </option>
          <option data-testid="option-sub" value={CalculatorOperation.SUB}>
            -
          </option>
          <option
            data-testid="option-multiply"
            value={CalculatorOperation.MULTIPLY}
          >
            x
          </option>
          <option data-testid="option-div" value={CalculatorOperation.DIVISION}>
            /
          </option>
        </select>
        <input
          data-testid="secondNumber"
          type="number"
          placeholder="Number2"
          value={props.secondNumber}
          onChange={
            props.onSecondNumberChange ? props.onSecondNumberChange : () => {}
          }
        />
        {props.hasError ? (
          <p data-testid="errorMessage" className={styles.error}>
            {props.hasError}
          </p>
        ) : (
          <input
            data-testid="result"
            type="number"
            placeholder="Result"
            disabled
            value={props.result}
            readOnly
          />
        )}
      </div>
    </div>
  );
}
