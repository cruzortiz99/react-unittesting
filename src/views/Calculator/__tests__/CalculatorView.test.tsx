import { fireEvent, render, screen } from "@testing-library/react"
import { CalculatorOperation, CalcultatorView, getCalculatorOperationFromString } from "../CalculatorView"

describe("Calculator View", () => {
  test("Should render as expected when the component is mounted", () => {
    const {container} = render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} />)
    expect(container).toMatchSnapshot()
  })
  test("Should show the value of firstNumber when is passed as props", async () => {
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} />)
    const firstNumberInput = await screen.findByTestId(/firstNumber/)
    expect(firstNumberInput.getAttribute('value')).toBe("10")
  })
  test("Should emit onFistNumberChange when first number input changes", async () => {
    const mockFn = jest.fn()
    render(<CalcultatorView firstNumber={10} secondNumber={10} operation={CalculatorOperation.ADD} result={20} onFirstNumberChange={mockFn} />)
    const firstNumberInput = await screen.findByTestId(/firstNumber/)
    fireEvent.change(firstNumberInput, {target: {value: '11'}})
    expect(mockFn).toHaveBeenCalledTimes(1)
  })
  test("Should show the value of secondNumber when is passed as props", async () => {
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} />)
    const secondNumberInput = await screen.findByTestId(/secondNumber/)
    expect(secondNumberInput.getAttribute('value')).toBe("15")
  })
  test("Should emit onSecondNumberChange when second number input changes", async () => {
    const mockFn = jest.fn()
    render(<CalcultatorView firstNumber={10} secondNumber={10} result={20} operation={CalculatorOperation.ADD} onSecondNumberChange={mockFn}/>)
    const secondInput = await screen.findByTestId(/secondNumber/)
    fireEvent.change(secondInput, {target: {value: "3"}})
    expect(mockFn).toHaveBeenCalledTimes(1)
  })
  test("Should show the value of result when is passed as props", async () => {
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} />)
    const resultNumber = await screen.findByTestId(/result/)
    expect(resultNumber.getAttribute('value')).toBe("90")
  })
  test("Should show message error when hasError is passed as props", async () =>{
    const testMessageError = "Test message error"
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} hasError={testMessageError} />)
    const messageError = await screen.findByTestId(/errorMessage/)
    expect(messageError.textContent).toBe(testMessageError)
  })
  test("Should hide result when hasError is passed as props", async () =>{
    const testMessageError = "Test message error"
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} hasError={testMessageError} />)
    const resultInput =  screen.queryByTestId(/result/)
    expect(resultInput).toBeFalsy()
  })
  test("Should show the correct value when the operation is passed as props", async () => {
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} />)
    const optionSelected: HTMLOptionElement = await screen.findByTestId(/option-multiply/)
    expect(optionSelected.selected).toBeTruthy()
    const optionUnselected: HTMLOptionElement = await screen.findByTestId(/option-add/)
    expect(optionUnselected.selected).toBeFalsy()
  })
  test("Should emit onOperationSelected when operation option changes", async () => {
    const mockFn = jest.fn()
    render(<CalcultatorView firstNumber={10} secondNumber={15} operation={CalculatorOperation.MULTIPLY} result={90} onOperationSelected={mockFn}/>)
    const selectOperation = await screen.findByTestId(/operation/)
    fireEvent.change(selectOperation, {target: { value: CalculatorOperation.SUB}})
    expect(mockFn).toHaveBeenCalledTimes(1)
    expect(mockFn).toHaveBeenCalledWith(CalculatorOperation.SUB)
  })
  test("Should get the right Calculator operation when the string is match", () => {
    Object.values(CalculatorOperation).forEach((value) => {
      const operation = getCalculatorOperationFromString(`${value}`.toLowerCase())
      expect(operation).toBe(value)
    })
  })
  test("Should get operation add when string does not match valid option", () => {
    const operation = getCalculatorOperationFromString("---")
    expect(operation).toBe(CalculatorOperation.ADD)
  })
})