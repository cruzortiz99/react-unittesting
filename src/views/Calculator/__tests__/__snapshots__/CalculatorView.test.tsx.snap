// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator View Should render as expected when the component is mounted 1`] = `
<div>
  <div>
    <h3>
      Calculator
    </h3>
    <div
      class="calculatorContainer"
    >
      <input
        data-testid="firstNumber"
        placeholder="Number1"
        type="number"
        value="10"
      />
      <select
        data-testid="operation"
      >
        <option
          data-testid="option-add"
          value="add"
        >
          +
        </option>
        <option
          data-testid="option-sub"
          value="sub"
        >
          -
        </option>
        <option
          data-testid="option-multiply"
          value="multiply"
        >
          x
        </option>
        <option
          data-testid="option-div"
          value="div"
        >
          /
        </option>
      </select>
      <input
        data-testid="secondNumber"
        placeholder="Number2"
        type="number"
        value="15"
      />
      <input
        data-testid="result"
        disabled=""
        placeholder="Result"
        readonly=""
        type="number"
        value="90"
      />
    </div>
  </div>
</div>
`;
